### DISCUSSION TITLE
OpenTelemtry integration and promise contexts

### DISCUSSION THREAD

**Discussion by Qudratillo** [activej/activej#264] [Q&A] [answered]
I want to use OpenTelemetry. I want to make as automated as possible. I need to store trace-id somewhere when there is `map` or `then` called in promises.
Is there anything like subscriber context of Project Reactor?

**Comment by Qudratillo** [comment]
Hi @eduard-vasinskyi,
Can you help me?

**Comment by eduard-vasinskyi** [comment, marked as answer]
Hi, @Qudratillo

We have designed ActiveJ promises to be as lightweight as possible. There is no built-in context for that reason. You can pass trace-id or any other context explicitly as a method parameter, for example. 

There is, however, a context inside an [HttpMessage](https://github.com/activej/activej/blob/af631c65e0ab496222898eb93ec7a0d4b898b5e5/core-http/src/main/java/io/activej/http/HttpMessage.java) (`HttpRequest` and `HttpResponse`). If your code uses `activej-http` you can pass required variables as `HttpMessage` attachments (see [HttpMessage#attach](https://github.com/activej/activej/blob/af631c65e0ab496222898eb93ec7a0d4b898b5e5/core-http/src/main/java/io/activej/http/HttpMessage.java#L301-L358), [HttpMessage#getAttachment](https://github.com/activej/activej/blob/af631c65e0ab496222898eb93ec7a0d4b898b5e5/core-http/src/main/java/io/activej/http/HttpMessage.java#L360-L406)). This is useful for passing context between different servlets, like routing servlet and some downstream servlets.

**Reply by Qudratillo** [reply]
Hi @eduard-vasinskyi ,

Thank you for a good explanation!

Could we have some kind of flag to enable Promise contexts? I believe contexts would decrease developer burden significantly with little performance cost. OpenTelemetry is a good example of such case.

**Reply by eduard-vasinskyi** [reply]
We are not planning to add context to Promises right now, but we will consider it for the next ActiveJ versions.

**Comment by kcsurapaneni** [comment]
@eduard-vasinskyi By default, the trace ID and span ID are not automatically attached to the logs in the ActiveJ application. To enable this functionality, do we need to configure it manually. We are using Slf4J, Logback, and the [logback-mdc-1.0](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/logback/logback-mdc-1.0) library, but it's not working out of the box. Could you provide guidance on how to set this up?

**Reply by eduard-vasinskyi** [reply]
Hi, @kcsurapaneni 
Answered [here](https://github.com/activej/activej/issues/338#issuecomment-2651609003)
